### DISCUSSION TITLE
relaxed load and store

### DISCUSSION THREAD

**Discussion by jinz2014** [NVIDIA/cccl#1573] [CUB]
Can you please provide a solution that is portable for relaxed load and store operations ? Currently, the assembly codes (e.g. load) are required.  Thanks.

```
static __device__ __forceinline__ uint4 load_relaxed(uint4 const *ptr)
{
#ifdef PTX
  uint4 retval;
  NV_IF_TARGET(NV_PROVIDES_SM_70,
               (asm volatile("ld.relaxed.gpu.v4.u32 {%0, %1, %2, %3}, [%4];"
                             : "=r"(retval.x), "=r"(retval.y), "=r"(retval.z), "=r"(retval.w)
                             : _CUB_ASM_PTR_(ptr)
                             : "memory");),
               (asm volatile("ld.cg.v4.u32 {%0, %1, %2, %3}, [%4];"
                             : "=r"(retval.x), "=r"(retval.y), "=r"(retval.z), "=r"(retval.w)
                             : _CUB_ASM_PTR_(ptr)
                             : "memory");));
  return retval;
#else
 return *ptr;
#endif
}
```

**Comment by miscco** [comment]
Could you provide some more details. What does portable mean in this context? Different architectures, other vendors?

**Reply by jinz2014** [reply]
Yes. A CUDA program without assembly codes for vendors.

**Comment by jrhemstad** [comment]
Using `cuda::atomic_ref::load/store(cuda::memory_order_relaxed)` is probably your easiest solution. 

https://godbolt.org/z/39cKqhnW3

**Reply by jinz2014** [reply]
Thank you for the link. However, I find that atomic_ref does not support types larger than 8 bytes.
